pkg/ddl/ddl_worker.go

```diff
@@ -899,7 +907,27 @@ func (w *worker) prepareTxn(job *model.Job) (kv.Transaction, error) {
 	return txn, err
 }
 
-func (w *worker) HandleDDLJobTable(d *ddlCtx, job *model.Job) (int64, error) {
+// runOneJobStep runs one step of the DDL job and persist the states change. One
+// *step* is defined as the following reason:
+//
+// - TiDB uses "Asynchronous Schema Change in F1", one job may have multiple
+// *steps* each for a "schema state" such as 'delete only' -> 'write only'. The
+// cluster can only have at most two adjacent schema states for a DDL object.
+// This function forwards one (schema state) *step* and caller
+// `runOneJobStepAndWaitSync` waits for other nodes to catch up.
+//
+// - In order to make sure every job is deterministic, idempotent to provide
+// correctness through failover, this function will decide and persist the
+// arguments of a job as a separate *step*. These steps will reuse "schema state"
+// changes, see onRecoverTable as an example.
+//
+// - We may need to use caller `runOneJobStepAndWaitSync` to make sure other node
+// is synchronized before change the job state. So an extra job state *step* is
+// added.
```

Comment:
you mean the wait change above runOneJobStep? it's for failover,

Reply:
Yes the wait change above runOneJobStep. This item wants to describe the job state changes, for example from JobStateDone to JobStateSynced. 

I'm not sure why JobStateDone -> JobStateSynced is relevent to failover. My current understanding is current node (the user connected node) must wait all other nodes to finished synchronize job state, otherwise it breaks the  linearizability (the user connected node shows DDL is finished, but the slow node shows DDL is not finished)

Reply:
`done`->`synced` might not go through `the wait change above runOneJobStep`, `wait` part might have done in below `waitSchemaChanged`

> JobStateDone -> JobStateSynced is relevant to failover.

suppose owner change during wait, with this state change, we can catch it and wait again on new owner, i.e. the `the wait change above runOneJobStep`